### PR TITLE
Add completion for binaries on PATH

### PR DESCRIPTION
**Pull request by AaronC81**
This PR adds completion for executables found on the user's PATH, just like how built-in commands such as `ls` and `version` are currently completed.

The `no_bin_complete` config option can be set to `$true` to disable this behaviour.

This also alters the behaviour of where certain autocompletion options are available, making commands and executables only appear as autocompletion options when the user is typing a command, and not when they are typing its arguments. For example, `echo ver<tab>` used to complete to `echo version` because of the existence of the `version` command, which isn't ideal; this kind of completion no longer occurs.

### CODE REVIEW

**Comment by thegedge** [comment]
Nice! 🎉

Looks like CI isn't reporting. Here's the failed build: https://dev.azure.com/nushell/nushell/_build/results?buildId=2869&view=logs&j=3df22538-2a64-5d6a-ccd1-435dcd87d567&t=2a625c33-ace7-5ee9-d25b-d3e450c793b3

In case you can't view it, these are failing on windows:

```
error[E0433]: failed to resolve: could not find `unix` in `os`
  --> crates\nu-cli\src\shell\completer.rs:10:14
   |
10 | use std::os::unix::fs::PermissionsExt;
   |              ^^^^ could not find `unix` in `os`

error[E0425]: cannot find function `pathext` in this scope
   --> crates\nu-cli\src\shell\completer.rs:223:24
    |
223 |             let exts = pathext()?;
    |                        ^^^^^^^ not found in this scope

error[E0433]: failed to resolve: could not find `unix` in `os`
  --> crates\nu-cli\src\shell\completer.rs:10:14
   |
10 | use std::os::unix::fs::PermissionsExt;
   |              ^^^^ could not find `unix` in `os`

error[E0425]: cannot find function `pathext` in this scope
   --> crates\nu-cli\src\shell\completer.rs:223:24
    |
223 |             let exts = pathext()?;
    |                        ^^^^^^^ not found in this scope

```

**Comment by AaronC81** [comment]
I added a [new commit](https://github.com/nushell/nushell/pull/1866/commits/a0efc25c8c75621b63166b84a1a0c25b372d03b1) since then which (hopefully) fixes that issue, but it doesn't seem to have re-run the CI :/

Looks like Azure DevOps fell over in one of these logs: https://dev.azure.com/nushell/nushell/_build/results?buildId=2869&view=logs&j=dc8d491f-9069-51dc-3b15-fee2dae0e5e2

Is there a way I can force DevOps to re-run, or should I make a tiny/blank commit?

**Comment by thegedge** [comment]
Ahh, I was looking at an old failing build. 

> Is there a way I can force DevOps to re-run, or should I make a tiny/blank commit?

I couldn't find a way to manually kick off a run, so what you can do is a simple `git commit --amend --no-edit && git push --force-with-lease`, which should hopefully get a new build started.

**Comment by AaronC81** [comment]
Thanks, that worked - looks like it still runs into some problems on Windows though. I'll get those sorted out!

**Comment by AaronC81** [comment]
Got this working on Windows now!

File: crates/nu-cli/src/shell/completer.rs [hunk 1]
```diff
@@ -173,4 +209,72 @@ impl NuCompleter {
...
+    fn get_path_executables(&self) -> Option<HashSet<String>> {
+        let path_var = std::env::var_os("PATH")?;
+        let paths: Vec<_> = std::env::split_paths(&path_var).collect();
+
+        let mut executables: HashSet<String> = HashSet::new();
+        for path in paths {
+            if let Ok(mut contents) = read_dir(path) {
+                while let Some(Ok(item)) = contents.next() {
+                    if let Ok(is_ex) = self.is_executable(&item) {
+                        if is_ex {
```
Comment (thegedge) [inline crates/nu-cli/src/shell/completer.rs:268]: You can avoid another level of indentation by writing

``` rust
if let Ok(true) = self.is_executable() {  // I think this works
```

or

```rust
if self.is_executable().unwrap_or(false) {
```

File: crates/nu-cli/src/shell/completer.rs [hunk 2]
```diff
@@ -173,4 +209,72 @@ impl NuCompleter {
...
+
+        let filetype = metadata.file_type();
+        let permissions = metadata.permissions();
+
+        // The file is executable if it is a directory or a symlink and the permissions are set for
+        // owner, group, or other
+        Ok((filetype.is_file() || filetype.is_symlink()) && (permissions.mode() & 0o111 != 0))
+    }
+
+    fn get_path_executables(&self) -> Option<HashSet<String>> {
```
Comment (thegedge) [inline crates/nu-cli/src/shell/completer.rs:259]: Naming tip that I've learned over the years: `get_` is almost never a useful prefix. Functions that return something almost always "get" you something 🙂

File: crates/nu-cli/src/shell/completer.rs [hunk 3]
```diff
@@ -56,27 +71,46 @@ impl NuCompleter {
...
+                _ => false,
+            },
+            _ => false,
+        };
+
+        // Only complete executables or commands if the thing we're completing
+        // is syntactically a command
+        if replace_loc == ReplacementLocation::Command {
+            let mut all_executables: HashSet<_> = commands.iter().map(|x| x.to_string()).collect();
+            if !no_bin_complete {
```
[the remark this answers is not in the document]
Comment (AaronC81) [reply crates/nu-cli/src/shell/completer.rs]: I was trying to follow the pattern of the existing names, but I agree that something positive is clearer.

File: crates/nu-cli/src/shell/completer.rs [hunk 4]
```diff
@@ -173,4 +209,72 @@ impl NuCompleter {
...
+        // The file is executable if it is a directory or a symlink and the permissions are set for
+        // owner, group, or other
+        Ok((filetype.is_file() || filetype.is_symlink()) && (permissions.mode() & 0o111 != 0))
+    }
+
+    fn get_path_executables(&self) -> Option<HashSet<String>> {
+        let path_var = std::env::var_os("PATH")?;
+        let paths: Vec<_> = std::env::split_paths(&path_var).collect();
+
+        let mut executables: HashSet<String> = HashSet::new();
```
[the remark this answers is not in the document]
Comment (AaronC81) [reply crates/nu-cli/src/shell/completer.rs]: Good idea!

File: crates/nu-cli/src/shell/completer.rs (hunk 1 above)
Comment (AaronC81) [reply crates/nu-cli/src/shell/completer.rs]: That first solution is brilliant, thanks

File: crates/nu-cli/src/shell/completer.rs (hunk 2 above)
Comment (AaronC81) [reply crates/nu-cli/src/shell/completer.rs]: True! `find` works better here I think.

**Comment by AaronC81** [comment]
Thanks for your review - I'm relatively new to Rust so it's really helpful.
I can't reproduce your issue I'm afraid :/

File: crates/nu-cli/src/shell/completer.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (thegedge) [reply crates/nu-cli/src/shell/completer.rs]: > I was trying to follow the pattern of the existing names

Totally fair. I chatted to the team about this, and moving forward we'll try to stick to names without negatives. We'll have to go back and change some existing configs though 🙂

**Comment by thegedge** [comment]
I'll give this another spin sometime today, and if all looks well, I'll get it merged 🙂

**Comment by thegedge** [comment]
I don't know what it is, but sometimes tab completion doesn't show me the list of possible options. That's a separate problem though, as command completion itself is working 🙌

**Pull request merged**